### ISSUE TITLE
Catch and handle ChunkLoadError

### ISSUE THREAD

**Issue by aaronadamsCA** [gatsbyjs/gatsby#33844]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

We continue to see unhandled `ChunkLoadError` reports in our Sentry logs, as previously tracked in #18866.

### Reproduction Link

N/A

### Steps to Reproduce

N/A

### Expected Result

Gatsby should catch `ChunkLoadError` and then examine and handle the error.

### Actual Result

I continue to see `ChunkLoadError` unhandled exceptions in our Sentry logs.

@KyleAMathews added a workaround in #33032, where the page reloads on `ChunkLoadError`, but issues remain:

- The error still bubbles to the app, and then to our Sentry logs.
- If a chunk really was missing, for example due to a Webpack or CDN issue, the application would enter into an infinite loop. In a worst-case scenario this could DDoS an already-broken server or host.
  - This can be reproduced by using DevTools to block requests for a single `.js` chunk and refreshing. The application starts aggressively refreshing.

I think the current workaround was written the way it is because @KyleAMathews believed this error is uncatchable: https://github.com/gatsbyjs/gatsby/issues/18866#issuecomment-845561804

But I'm reasonably confident that I wouldn't be seeing this in my Sentry logs if that were true. If our error boundary can catch it, certainly a Gatsby error boundary could do the same.

With some guidance I'd be happy to take a shot at a PR. Specifically I'd love any ideas on avoiding the refresh loop - I think it would be better to crash on a second consecutive error - and on whether it would be possible to write tests that cover this.

### Environment

```shell
Gatsby latest.
```

**Comment by graysonhicks** [comment]
Hi @aaronadamsCA !  From the errors that have made it to your Sentry, any ideas on a way to create a reproduction for debugging and testing?

**Comment by aaronadamsCA** [comment]
@graysonhicks Easiest way I've found:

> This can be reproduced by using DevTools to block requests for a single `.js` chunk and refreshing. The application starts aggressively refreshing.

Blocking the chunk that contains `siteMetadata` has been a pretty reliable means of reproduction, and you can do this on any existing Gatsby site via DevTools.

That said, I assume a programmatic reproduction would be needed in order to make this truly testable.

I don't know much about Gatsby's Webpack internals, but could we somehow create a minimal repro site with a haunted Webpack config that deletes its own `siteMetadata` chunk?

**Comment by pieh** [comment]
Seems like we do catch those errors and then specifically rethrow them in  https://github.com/gatsbyjs/gatsby/blob/070eeee4bf80921dc50f376369683f03ed399824/packages/gatsby/cache-dir/production-app.js#L165-L172

We can add some specific handling for `ChunkLoadError` errors, but just silencing them doesn't seem good idea. Not sure what special handling of this mean exactly.

Here's me reproducing the error:
![Screenshot 2021-11-29 at 19 36 38](https://user-images.githubusercontent.com/419821/143923937-f29ce9de-aa7b-483b-9f53-a3c9abdcd446.png)

**Comment by aaronadamsCA** [comment]
Yeah, I'm noticing a significant rise in this error in Sentry now:

`We couldn't load "/page-data/sq/d/{number}.json"`

Which also appears to be thrown by that same line of code you just highlighted, @pieh.

From what I see, the error happens because a JSON file that exists on the server fails to load (status code 0 instead of 200). This, then, is also probably a transient condition that would clear up on reload, just like `ChunkLoadError`.

From the outside looking in, then, it seems like a good simple strategy here could be:

- **For this issue:** Check if `page.error.name === "ChunkLoadError"`, if yes, issue `window.location.reload()` instead of throwing.
  - Also remove the `onunhandledrejection` handler, since it would no longer be needed.
- **For each similar issue:** Update the error to use its own custom error class, then expand the `page.error.name` check in `production-app.js` to match the additional error name.

**Comment by PeterDekkers** [comment]
> * **For this issue:** Check if `page.error.name === "ChunkLoadError"`, if yes, issue `window.location.reload()` instead of throwing.

@aaronadamsCA, if the chunk is actually missing then wouldn't this turn into a reload loop?

**Comment by graysonhicks** [comment]
@PeterDekkers Yea, I think he mentioned that as the worst-case scenario in the description.  Definitely something that would need to be solved.

**Comment by PeterDekkers** [comment]
Ah yes, thank you for clarifying @graysonhicks.

**Comment by valentinfily** [comment]
Any update on this issue ? Thanks

**Comment by LeonxLJX** [comment]
I'd like to take this one (`Catch and handle ChunkLoadError`). I'll look into the root cause and follow up with a PR. (claiming via @LeonxLJX)

**Comment by LeonxLJX** [comment]
I'd like to take this one (`Catch and handle ChunkLoadError`). I'll look into the root cause and follow up with a PR. (claiming via @LeonxLJX)
